builder/builder.go

```diff
@@ -0,0 +1,386 @@
+	var s string
+
+	for i := range v.keys {
+		switch v := v.keys[i].(type) {
+		case string:
+			s += v + " "
+		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+			s += fmt.Sprintf("%d ", v)
+		case float32, float64:
+			s += fmt.Sprintf("%f ", v)
+		default:
+			panic(fmt.Sprintf("unsupported key type: %T", v))
+		}
+	}
+
+	s += "{{"
+
+	for i := range v.pattern {
+		if i != 0 {
+			s += " "
+		}
+
+		switch v := v.pattern[i].(type) {
+		case string:
+			s += v
+		case *Expression:
+			s += v.String()
+		default:
+			panic(fmt.Sprintf("unsupported pattern type: %T", v))
+		}
+	}
+
+	return s + "}}"
+}
+
+type local struct {
+	expr     *Expression
+	variable variable
+}
+
+type literal any
+
+type variable string
+
+type function struct {
+	name    string
+	options []option
+}
+
+type Expression struct {
+	spacing  string
+	operand  any // literal or variable
+	function function
+}
+
+func (e *Expression) String() string {
+	s := "{" + e.spacing
+
+	switch v := e.operand.(type) {
+	case variable:
+		s += string(v)
+	case nil:
+		// noop
+	case literal:
+		s += printLiteral(v)
+	default:
+		panic(fmt.Sprintf("unsupported operand type: %T", v))
+	}
+
+	if e.function.name != "" {
+		if e.operand != nil { // literal or variable
+			s += " "
+		}
+
+		s += e.function.name
+
+		for _, o := range e.function.options {
+			s += " " + o.key + e.spacing + "=" + e.spacing + fmt.Sprint(o.operand)
+		}
+	}
+
+	return s + e.spacing + "}"
+}
+
+func Expr() *Expression { return new(Expression) }
+
+func (e *Expression) Literal(v any) *Expression {
+	e.operand = v
+	return e
+}
+
+func (e *Expression) Var(v string) *Expression {
+	if len(v) == 0 {
+		panic("variable name cannot be empty")
+	}
+
+	if v[0] != '$' {
+		panic(fmt.Sprintf("variable must start with $: %s", v))
+	}
+
+	e.operand = variable(v)
+
+	return e
+}
+
+type option struct { //nolint:govet
+	key     string
+	operand any // literal or variable
+}
+
+func Option(key string, operand any) option { return option{key, operand} }
+
+func (e *Expression) Func(name string, option ...option) *Expression {
+	if len(name) == 0 {
+		panic("function name cannot be empty")
+	}
+
+	if e.function.name != "" {
+		panic("expression already has a function")
+	}
+
+	switch name[0] {
+	default:
+		panic(fmt.Sprintf("function MUST start with :, + or -: %s", name))
+	case ':', '+', '-':
+		e.function = function{name: name, options: option}
+		return e
+	}
+}
+
+func printLiteral(l any) string {
+	switch v := l.(type) {
+	case string:
+		v = quotedEscape(v)
+		return "|" + v + "|"
+	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+		return fmt.Sprintf("|%d|", v)
+	case float32, float64:
+		return fmt.Sprintf("|%f|", v)
+	default:
+		panic(fmt.Sprintf("unsupported literal type: %T", v))
+	}
+}
+
+// helpers
+
+// quotedEscape escapes special characters in quoted name literal.
+func quotedEscape(s string) string {
+	var sb strings.Builder
```

Comment:
strings.Replacer is cleaner and more straigforward

Reply:
is it going to be faster?

Reply:
`readability > speed` almost in any scenario

Reply:
<img width="874" alt="image" src="https://github.com/expect-digital/go-mf2/assets/69954124/35c6f46f-2cd5-483c-a9a3-959cbc5ab131">

strings.Replacer is faster

Reply:
@VladislavsPerkanuks can you add benchtest for inline string concat?